File: Term.java
import java.util.ArrayList;

public class Term {
    private static int[] ALPHA_VALUE_EXPONENT = new int[257];
    private final static int[] ALPHA_VALUE = new int[257];
    private final static String[] SUPERSCRIPT = { "\u2070", "\u00B9", "\u00B2", "\u00b3", "\u2074", "\u2075", "\u2076",
            "\u2077", "\u2078", "\u2079" };
    // x refers to the exponent of x
    private int xExponent;
    // alpha refers to the exponent of α
    private int alphaExponent;
    static {
        for (int i = 0; i < 255; i++) {
            int alpha = getAlpha(i);
            ALPHA_VALUE_EXPONENT[alpha] = i;
            ALPHA_VALUE[i] = alpha;
        }
        ALPHA_VALUE[255] = 1;
        ALPHA_VALUE_EXPONENT[0] = 256;
    }

    public Term(int xExponent, int alphaExponent) {
        this.xExponent = xExponent;
        this.alphaExponent = alphaExponent;
    }

    public Term multiplyTerms(Term term) {
        return new Term(xExponent + term.xExponent, (alphaExponent + term.alphaExponent) % 255);
    }

    public static Term[] multiplyPolynomials(Term[] poly1, Term[] poly2) {
        Term[] result = new Term[poly1.length * poly2.length];
        for (int i = 0; i < poly1.length; i++)
            for (int j = 0; j < poly2.length; j++)
                result[i * poly2.length + j] = poly1[i].multiplyTerms(poly2[j]);
        return result;
    }

    public static Term[] multiplyX(Term[] polynomial, int x) {
        Term[] result = new Term[polynomial.length];
        for (int i = 0; i < polynomial.length; i++)
            result[i] = new Term(polynomial[i].xExponent + x, polynomial[i].alphaExponent);
        return result;
    }

    public static Term[] multiplyAlpha(Term[] polynomial, int alpha) {
        Term[] result = new Term[polynomial.length];
        for (int i = 0; i < polynomial.length; i++)
            result[i] = new Term(polynomial[i].xExponent, (polynomial[i].alphaExponent + alpha) % 255);
        return result;
    }

    public static Term[] xor(Term[] poly1, Term[] poly2) {
        int limit = Math.min(poly1.length, poly2.length);
        int max = Math.max(poly1.length, poly2.length);
        Term[] result = new Term[max - 1];
        int i = 0;
        for (; i < limit - 1; i++)
            result[i] = new Term(poly1[i + 1].xExponent, ALPHA_VALUE_EXPONENT[ALPHA_VALUE[poly1[i + 1].alphaExponent]
                    ^ ALPHA_VALUE[poly2[i + 1].alphaExponent]]);
        if (poly1.length > poly2.length)
            for (; i < result.length; i++)
                result[i] = poly1[i + 1].clone();
        else
            for (; i < result.length; i++)
                result[i] = poly2[i + 1].clone();
        return result;
    }

    public Term clone() {
        return new Term(xExponent, alphaExponent);
    }

    public Term add(Term term) {
        int alpha1 = ALPHA_VALUE[this.alphaExponent];
        int alpha2 = ALPHA_VALUE[term.alphaExponent];
        return new Term(xExponent, ALPHA_VALUE_EXPONENT[alpha1 ^ alpha2]);
    }

    public static Term[] simplify(Term[] polynomial, int difference) {
        Term[] list = new Term[polynomial.length - difference];
        int xExponent = polynomial[0].xExponent;
        int index = 0;
        Term tally = polynomial[0];
        for (int i = 1; i < polynomial.length; i++)
            if (polynomial[i].xExponent != xExponent) {
                list[index++] = tally;
                tally = polynomial[i];
                xExponent = polynomial[i].xExponent;
            } else
                tally = tally.add(polynomial[i]);
        list[index] = tally;
        return list;
    }

    public static int getAlpha(int exponent) {
        int result = 1;
        for (int i = 0; i < exponent; i++) {
            result *= 2;
            if (result > 255)
                result ^= 285;
        }
        return result;
    }

    public static String prettyPrint(Term... polynomial) {
        String result = "";
        for (int i = 0; i < polynomial.length; i++)
            result += "α" + toSuperscript(polynomial[i].alphaExponent) + "x" + toSuperscript(polynomial[i].xExponent)
                    + " + ";
        return result.substring(0, result.length() - 3);
    }

    public static String condensedPrettyPrint(Term... polynomial) {
        String result = "";
        for (int i = 0; i < polynomial.length; i++) {
            String alpha = polynomial[i].alphaExponent == 0 ? "" : "α" + toSuperscript(polynomial[i].alphaExponent);
            String x = polynomial[i].xExponent == 0 ? "" : "x" + toSuperscript(polynomial[i].xExponent);
            result += alpha + x + " + ";
        }
        return result.substring(0, result.length() - 3);
    }

    public static String NumericPrettyPrint(Term... polynomial) {
        String result = "";
        for (int i = 0; i < polynomial.length; i++) {
            String alpha = polynomial[i].alphaExponent == 0 ? "" : "" + ALPHA_VALUE[polynomial[i].alphaExponent];
            String x = polynomial[i].xExponent == 0 ? "" : "x" + toSuperscript(polynomial[i].xExponent);
            result += alpha + x + " + ";
        }
        return result.substring(0, result.length() - 3);
    }

    public static Term[] toPolynomial(int[] message) {
        Term[] polynomial = new Term[message.length / 8];
        for (int i = 0; i < message.length; i += 8) {
            int codeword = 0;
            for (int j = 0; j < 8; j++) {
                int bit = message[i + j] % 2;
                codeword |= (bit << j);
            }
            polynomial[polynomial.length - i / 8 - 1] = new Term(i / 8, ALPHA_VALUE_EXPONENT[codeword]);
        }
        return polynomial;
    }

    public static Term[] toPolynomialTemp(int[] message) {
        Term[] polynomial = new Term[message.length];
        for (int i = 0; i < polynomial.length; i++)
            polynomial[i] = new Term(polynomial.length - i - 1, ALPHA_VALUE_EXPONENT[message[i]]);
        return polynomial;
    }

    private static String toSuperscript(int number) {
        String result = "";
        int total = number;
        int m = 10;
        do {
            result = SUPERSCRIPT[total % m] + result;
            total /= 10;
        } while (total != 0);
        return result;
    }

    public int getAlphaExponent() {
        return alphaExponent;
    }

    public int getXExponent() {
        return xExponent;
    }

    public static int[] toErrorCodes(Term[] polynomial) {
        int[] message = new int[polynomial.length * 8];
        for (int i = 0; i < polynomial.length; i++) {
            int codeword = ALPHA_VALUE[polynomial[i].alphaExponent];
            for (int j = 7; j > -1; j--) {
                message[message.length - 1 - i * 8 - j] = 2 + (codeword % 2);
                codeword >>>= 1;
            }
        }
        return message;
    }

    public static int[] toArray(Term[] polynomial) {
        int[] result = new int[polynomial.length];
        for (int i = 0; i < polynomial.length; i++)
            result[i] = ALPHA_VALUE[polynomial[i].alphaExponent];
        return result;
    }

    public static Term[] genPolynomial(int errorCodeCount) {
        // index 0 == exponent of x
        // index 1 == α
        Term[][] terms = new Term[errorCodeCount][2];
        for (int i = 0; i < errorCodeCount; i++) {
            terms[i][0] = new Term(1, 0);
            terms[i][1] = new Term(0, i);
        }
        Term[] result = terms[0];
        for (int i = 1; i < errorCodeCount; i++) {
            result = Term.multiplyPolynomials(result, terms[i]);
            result = Term.simplify(result, i);
        }
        return result;
    }
}

File: src/Term.cpp
#include "Term.hpp"

int get_alpha(int exponent)
{
    int result = 1;
    for (int i = 0; i < exponent; i++)
    {
        result *= 2;
        if (result > 255)
            result ^= 285;
    }
    return result;
}

void multiply_x(Polynomial &result, Polynomial &p, int x)
{
    for (int i = 0; i < p.length; i++){
        result[i].xExponent = p[i].xExponent + x;
        result[i].alphaExponent = p[i].alphaExponent;
    }
}

void multiply_alpha(Polynomial &result, Polynomial &p, int alpha)
{
    for (int i = 0; i < p.length; i++){
        result[i].xExponent = p[i].xExponent;
        result[i].alphaExponent = (p[i].alphaExponent + alpha) % 255;
    }
}

Polynomial simplify(Polynomial &p, int difference)
{
    Polynomial list(p.length - difference);
    int xExponent = p[0].xExponent;
    int index = 0;
    Term tally = Term();
    tally.xExponent = p[0].xExponent;
    tally.alphaExponent = p[0].alphaExponent;
    for (int i = 1; i < p.length; i++)
        if (p[i].xExponent != xExponent)
        {
            list[index].xExponent = tally.xExponent;
            list[index++].alphaExponent = tally.alphaExponent;
            tally.xExponent = p[i].xExponent;
            tally.alphaExponent = p[i].alphaExponent;
            xExponent = p[i].xExponent;
        }
        else{
            Term tmp = tally + p[i];
            tally.xExponent =  tmp.xExponent;
            tally.alphaExponent = tmp.alphaExponent;
        }
    list[index].xExponent = tally.xExponent;
    list[index].alphaExponent = tally.alphaExponent;
    return list;
}

Polynomial to_polynomial(int *message, int len)
{
    Polynomial polynomial(len / 8);
    for (int i = 0; i < len; i += 8)
    {
        int codeword = 0;
        for (int j = 0; j < 8; j++)
        {
            int bit = message[i + j] % 2;
            codeword |= (bit << j);
        }
        polynomial[(len - i) / 8 - 1] = Term(i / 8, ALPHA_VALUE_EXPONENT[codeword]);
    }
    return polynomial;
}

Polynomial to_polynomial_temp(int *message, int len)
{
    int length = (int)len;
    Polynomial polynomial(len);
    for (int i = 0; i < length; i++)
    {
        polynomial[i].xExponent  = length - i - 1;
        polynomial[i].alphaExponent = ALPHA_VALUE_EXPONENT[message[i]];
    }
    return polynomial;
}

int *to_error_codes(Polynomial &polynomial)
{
    int *message = new int[polynomial.length * 8];
    for (int i = 0; i < polynomial.length; i++)
    {
        int codeword = ALPHA_VALUE[polynomial[i].alphaExponent];
        for (int j = 7; j > -1; j--)
        {
            message[(polynomial.length - i) * 8 - 1 - j] = 2 + (codeword % 2);
            codeword >>= 1;
        }
    }
    return message;
}

void to_array(Polynomial &polynomial, int *&result)
{
    result = new int[polynomial.length];
    for (int i = 0; i < polynomial.length; i++)
        result[i] = ALPHA_VALUE[polynomial[i].alphaExponent];
}

Polynomial gen_polynomial(int errorCodeCount)
{
    Term **matrix = new Term *[errorCodeCount];
    for (int i = 0; i < errorCodeCount; i++)
        matrix[i] = new Term[2];
    for (int i = 0; i < errorCodeCount; i++)
    {
        Term t(1, 0);
        matrix[i][0] = t;
        Term t2(0, i);
        matrix[i][1] = t2;
    }
    Polynomial result = Polynomial(2);
    for (int i = 0; i < 2; i++)
    {
        result[i].xExponent = matrix[0][i].xExponent;
        result[i].alphaExponent = matrix[0][i].alphaExponent;
    }
    for (int i = 1; i < errorCodeCount; i++)
    {
        Polynomial tmp = Polynomial(2);
        for (int j = 0; j < 2; j++)
        {
            tmp[j].xExponent = matrix[i][j].xExponent;
            tmp[j].alphaExponent = matrix[i][j].alphaExponent;
        }
        result = (result * tmp);
        result = simplify(result, i);
    }

    for (int i = 0; i < errorCodeCount; i++)
        delete[] matrix[i];
    delete[] matrix;
    return result;
}

void pretty_print(Polynomial &p)
{
    char *result = new char[p.length * 100]();
    for (int i = 0; i < p.length; i++)
    {
        std::strcat(result, "x");
        std::sprintf(result + std::strlen(result), "%d", p[i].xExponent);
        std::strcat(result, "α");
        std::sprintf(result + std::strlen(result), "%d", p[i].alphaExponent);
        std::strcat(result, i < p.length - 1 ? " + " : "");
    }
    printf("%s\n", result);
    delete[] result;
}

void pretty_print(Term *t, int len){
    Polynomial p = Polynomial(len);
    p.terms = t;
    pretty_print(p);
}
